File: packages/java/src/main/java/fbc/Seguranca/Ataques/AtaqueIntegral.java
package fbc.Seguranca.Ataques;

import fbc.Seguranca.AlvoCriptografico;
import fbc.Seguranca.AtaqueInterface;
import fbc.Seguranca.ResultadoAtaque;
import fbc.Seguranca.Severidade;
import fbc.Seguranca.SkipAtaqueException;
import fbc.Seguranca.Util;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

/**
 * Ataque integral (square): fixa a chave e percorre TODOS os 256 valores de
 * um byte do IV (e da chave), fazendo XOR de todos os keystreams resultantes.
 *
 * Numa função aleatória, o XOR de 256 saídas é uniforme, então cada byte da
 * soma é zero com probabilidade 1/256. Se a cifra tiver difusão incompleta,
 * a saída como função do byte variado fica "quase bijetiva" e a soma tende a
 * zero MUITO mais que o acaso - um distinguisher clássico.
 *
 * Como uma única (chave, IV) tem variância alta, acumula várias tentativas
 * para estimar o viés de forma estável (comparado a p=1/256).
 */
public class AtaqueIntegral implements AtaqueInterface {
    private final int tamanhoBloco;
    private final int posicoesIv;
    private final int posicoesChave;
    private final int tentativas;
    private final double limiteZ;

    public AtaqueIntegral() {
        this(32, 8, 8, 16, 5.0);
    }

    public AtaqueIntegral(
            int tamanhoBloco, int posicoesIv, int posicoesChave, int tentativas, double limiteZ) {
        this.tamanhoBloco = tamanhoBloco;
        this.posicoesIv = posicoesIv;
        this.posicoesChave = posicoesChave;
        this.tentativas = tentativas;
        this.limiteZ = limiteZ;
    }

    @Override
    public String nome() {
        return "Integral (soma balanceada variando 1 byte de IV/chave)";
    }

    @Override
    public ResultadoAtaque executar(AlvoCriptografico alvo) {
        int tamIv = alvo.tamanhoIv();
        int tamChave = alvo.chaveDeTeste().getBytes(StandardCharsets.UTF_8).length;

        byte[] primeiro = alvo.gerarKeystreamBruto(
                alvo.chaveDeTeste().getBytes(StandardCharsets.UTF_8),
                Util.randomBytes(tamIv),
                "enc",
                this.tamanhoBloco);
        if (primeiro == null) {
            throw new SkipAtaqueException("Alvo não expõe gerarKeystreamBruto.");
        }

        List<String> distinguidores = new ArrayList<>();
        long zerosTotal = 0;
        long bytesTotal = 0;

        for (int t = 0; t < this.tentativas; t++) {
            byte[] chave = Util.randomBytes(tamChave);
            byte[] ivBase = Util.randomBytes(tamIv);

            for (int pos = 0; pos < Math.min(this.posicoesIv, tamIv); pos++) {
                long[] res = this.somarVariando(alvo, chave, ivBase, "iv", pos);
                zerosTotal += res[0];
                bytesTotal += this.tamanhoBloco;
                if (res[1] == 1) {
                    distinguidores.add("iv[" + pos + "]");
                }
            }

            for (int pos = 0; pos < Math.min(this.posicoesChave, tamChave); pos++) {
                long[] res = this.somarVariando(alvo, chave, ivBase, "key", pos);
                zerosTotal += res[0];
                bytesTotal += this.tamanhoBloco;
                if (res[1] == 1) {
                    distinguidores.add("key[" + pos + "]");
                }
            }
        }

        double p = 1.0 / 256;
        double esperado = bytesTotal * p;
        double desvio = Math.sqrt(bytesTotal * p * (1 - p));
        double z = desvio > 0 ? (zerosTotal - esperado) / desvio : 0.0;

        boolean vulneravel = !distinguidores.isEmpty() || z > this.limiteZ;

        if (!distinguidores.isEmpty()) {
            return new ResultadoAtaque(
                    nome(),
                    true,
                    Severidade.CRITICA,
                    "Soma balanceada (XOR zero) encontrada variando: "
                            + String.join(", ", distinguidores.subList(0, Math.min(10, distinguidores.size()))),
                    Map.of("distinguidores", distinguidores));
        }

        Map<String, Object> dados = new HashMap<>();
        dados.put("zeros", zerosTotal);
        dados.put("esperado", esperado);
        dados.put("z", z);

        return new ResultadoAtaque(
                nome(),
                vulneravel,
                vulneravel ? Severidade.MEDIA : Severidade.INFO,
                String.format(
                        Locale.ROOT,
                        "bytes de saída zerados=%d (esperado ~%.1f, z=%.2f) em %d amostras; limite z=%.1f",
                        zerosTotal,
                        esperado,
                        z,
                        bytesTotal,
                        this.limiteZ),
                dados);
    }

    /** @return [quantidade de bytes zerados, 1 se todos os bytes são zero, senão 0] */
    private long[] somarVariando(
            AlvoCriptografico alvo, byte[] chave, byte[] ivBase, String campo, int pos) {
        byte[] xor = new byte[this.tamanhoBloco];
        for (int v = 0; v < 256; v++) {
            byte[] k = chave.clone();
            byte[] iv = ivBase.clone();
            if (campo.equals("iv")) {
                iv[pos] = (byte) v;
            } else {
                k[pos] = (byte) v;
            }
            byte[] ks = alvo.gerarKeystreamBruto(k, iv, "enc", this.tamanhoBloco);
            xor = this.xorBytes(xor, ks);
        }

        long zeros = 0;
        boolean todosZeros = true;
        for (byte b : xor) {
            if (b == 0) {
                zeros++;
            } else {
                todosZeros = false;
            }
        }
        return new long[] {zeros, todosZeros ? 1 : 0};
    }

    private byte[] xorBytes(byte[] a, byte[] b) {
        byte[] out = new byte[a.length];
        for (int i = 0; i < a.length; i++) {
            out[i] = (byte) ((a[i] ^ b[i]) & 0xFF);
        }
        return out;
    }
}

File: packages/go/internal/seguranca/ataques/ataque_integral.go
package ataques

import (
	"fmt"
	"math"
	"strings"

	"criptografia/internal/seguranca"
)

// OpcoesIntegral configura o AtaqueIntegral.
type OpcoesIntegral struct {
	TamanhoBloco  int
	PosicoesIv    int
	PosicoesChave int
	Tentativas    int
	LimiteZ       float64
}

// AtaqueIntegral (square) fixa a chave e percorre todos os 256 valores de um
// byte do IV (e da chave), fazendo XOR de todos os keystreams resultantes. Numa
// função aleatória cada byte da soma é zero com probabilidade 1/256; difusão
// incompleta deixa a soma tender a zero muito mais que o acaso.
type AtaqueIntegral struct {
	tamanhoBloco  int
	posicoesIv    int
	posicoesChave int
	tentativas    int
	limiteZ       float64
}

// NovoAtaqueIntegral cria o ataque. Sem argumento, usa tamanhoBloco = 32,
// posicoesIv = 8, posicoesChave = 8, tentativas = 16 e limiteZ = 5.0.
func NovoAtaqueIntegral(opcoes ...OpcoesIntegral) *AtaqueIntegral {
	o := OpcoesIntegral{
		TamanhoBloco:  32,
		PosicoesIv:    8,
		PosicoesChave: 8,
		Tentativas:    16,
		LimiteZ:       5.0,
	}
	if len(opcoes) > 0 {
		o = opcoes[0]
	}
	return &AtaqueIntegral{
		tamanhoBloco:  o.TamanhoBloco,
		posicoesIv:    o.PosicoesIv,
		posicoesChave: o.PosicoesChave,
		tentativas:    o.Tentativas,
		limiteZ:       o.LimiteZ,
	}
}

// Nome devolve o nome exibido no relatório.
func (a *AtaqueIntegral) Nome() string {
	return "Integral (soma balanceada variando 1 byte de IV/chave)"
}

// Executar roda o ataque.
func (a *AtaqueIntegral) Executar(alvo seguranca.AlvoCriptografico) (seguranca.ResultadoAtaque, error) {
	tamIv := alvo.TamanhoIv()
	tamChave := len(alvo.ChaveDeTeste())

	distinguidores := []string{}
	zerosTotal := 0
	bytesTotal := 0

	for t := 0; t < a.tentativas; t++ {
		chave := seguranca.RandomBytes(tamChave)
		ivBase := seguranca.RandomBytes(tamIv)

		for pos := 0; pos < min(a.posicoesIv, tamIv); pos++ {
			xor, zeros := a.somarVariando(alvo, chave, ivBase, "iv", pos)
			zerosTotal += zeros
			bytesTotal += a.tamanhoBloco
			if a.todosZeros(xor) {
				distinguidores = append(distinguidores, fmt.Sprintf("iv[%d]", pos))
			}
		}

		for pos := 0; pos < min(a.posicoesChave, tamChave); pos++ {
			xor, zeros := a.somarVariando(alvo, chave, ivBase, "key", pos)
			zerosTotal += zeros
			bytesTotal += a.tamanhoBloco
			if a.todosZeros(xor) {
				distinguidores = append(distinguidores, fmt.Sprintf("key[%d]", pos))
			}
		}
	}

	p := 1.0 / 256.0
	esperado := float64(bytesTotal) * p
	desvio := math.Sqrt(float64(bytesTotal) * p * (1 - p))
	z := 0.0
	if desvio > 0 {
		z = (float64(zerosTotal) - esperado) / desvio
	}

	vulneravel := len(distinguidores) > 0 || z > a.limiteZ

	if len(distinguidores) > 0 {
		limite := min(10, len(distinguidores))
		return seguranca.ResultadoAtaque{
			NomeAtaque: a.Nome(),
			Vulneravel: true,
			Severidade: seguranca.SevCritica,
			Detalhes:   "Soma balanceada (XOR zero) encontrada variando: " + strings.Join(distinguidores[:limite], ", "),
			Dados:      map[string]any{"distinguidores": distinguidores},
		}, nil
	}

	severidade := seguranca.SevInfo
	if vulneravel {
		severidade = seguranca.SevMedia
	}

	return seguranca.ResultadoAtaque{
		NomeAtaque: a.Nome(),
		Vulneravel: vulneravel,
		Severidade: severidade,
		Detalhes: fmt.Sprintf(
			"bytes de saída zerados=%d (esperado ~%.1f, z=%.2f) em %d amostras; limite z=%.1f",
			zerosTotal, esperado, z, bytesTotal, a.limiteZ,
		),
		Dados: map[string]any{"zeros": zerosTotal, "esperado": esperado, "z": z},
	}, nil
}

// somarVariando devolve o XOR de 256 keystreams e quantos bytes deram zero.
func (a *AtaqueIntegral) somarVariando(alvo seguranca.AlvoCriptografico, chave, ivBase []byte, campo string, pos int) ([]byte, int) {
	xor := make([]byte, a.tamanhoBloco)
	for v := 0; v < 256; v++ {
		k := append([]byte(nil), chave...)
		iv := append([]byte(nil), ivBase...)
		if campo == "iv" {
			iv[pos] = byte(v)
		} else {
			k[pos] = byte(v)
		}
		ks := alvo.GerarKeystreamBruto(k, iv, "enc", a.tamanhoBloco)
		xor = a.xorBytes(xor, ks)
	}

	zeros := 0
	for _, b := range xor {
		if b == 0 {
			zeros++
		}
	}
	return xor, zeros
}

// todosZeros informa se o buffer é todo de zeros.
func (a *AtaqueIntegral) todosZeros(buf []byte) bool {
	for _, b := range buf {
		if b != 0 {
			return false
		}
	}
	return true
}

// xorBytes faz XOR byte a byte de dois buffers.
func (a *AtaqueIntegral) xorBytes(a2, b []byte) []byte {
	out := make([]byte, len(a2))
	for i := 0; i < len(a2); i++ {
		out[i] = a2[i] ^ b[i]
	}
	return out
}
